File: 2015/day11.py
from typing import Callable
from os.path import isfile, join as path_join

file_name = path_join('input', 'day11.txt')


def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]


def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)


if not isfile(file_name):
	from aoc import get_input_file

	get_input_file()

old_password = open(file_name).read().strip()

inc = {"a": "b", "b": "c", "c": "d", "d": "e", "e": "f", "f": "g", "g": "h", "h": "i", "i": "j", "j": "k",
       "k": "l", "l": "m", "m": "n", "n": "o", "o": "p", "p": "q", "q": "r", "r": "s", "s": "t", "t": "u",
       "u": "v", "v": "w", "w": "x", "x": "y", "y": "z", "z": "a"}


def next(password, size):
	if password[-1] == 'z':  # overflow
		i = size - 1
		overflow = True
		while overflow and i >= 0:
			password[i] = 'a'
			i -= 1
			overflow = (password[i] == 'z')
			if not overflow:
				password[i] = inc[password[i]]
	else:
		password[-1] = inc[password[-1]]

	return password


def part1(password):
	new_pass = list(password)
	size = len(password)

	while True:
		new_pass = next(new_pass, size)

		# not contain i, o, l
		if "i" in new_pass or "o" in new_pass or "l" in new_pass:
			continue

		# increasing straight of 3 letters
		has_straight = False
		for i in range(size - 2):
			c1 = new_pass[i]
			c2 = inc[c1]
			c3 = inc[c2]

			if new_pass[i + 1] == c2 and new_pass[i + 2] == c3:
				has_straight = True
				break

		if not has_straight:
			continue

		# 2 different pairs
		def check_pairs():
			has_pairs = False
			for i in range(size - 2):
				if new_pass[i] == new_pass[i + 1]:
					pair1 = new_pass[i]
					for j in range(i + 2, size - 1):
						if new_pass[j] != pair1 and new_pass[j] == new_pass[j + 1]:
							return True
			return False

		if not check_pairs():
			continue

		return "".join(new_pass)


def part2(p):
	return part1(p)


p = part1(old_password)
print(p)
print(part2(p))

File: cpp/2015/day11.cpp
#include "../aocHelper.h"

class Day11 : public BaseDay
{
public:
	Day11() : BaseDay("11")
	{}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		auto get_next_string = [](char* text, int length)
		{
			if (text[length - 1] == 'z') // overflow
			{
				int i = length - 1;
				bool overflow = true;

				while (overflow && i >= 0)
				{
					text[i] = 'a';
					i--;
					overflow = (text[i] == 'z');

					if (!overflow)
					{
						text[i]++;
					}
				}
			}
			else // increment last char
			{
				text[length - 1]++;
			}
		};

		auto get_next_password = [&get_next_string](char* text, int length)
		{
			while (true)
			{
				get_next_string(text, length);

				// not contain i, o, l
				for (int i = 0; i < length; i++)
				{
					if (text[i] == 'i' || text[i] == 'l' || text[i] == 'l')
					{
						continue;
					}
				}

				// increasing straight of 3 letters
				bool has_straight = false;
				for (int i = 0; i < length - 2; i++)
				{
					char c1 = text[i];
					char c2 = text[i] + 1;
					char c3 = text[i] + 2;

					if (text[i + 1] == c2 && text[i + 2] == c3)
					{
						has_straight = true;
						break;
					}
				}

				if (!has_straight)
				{
					continue;
				}

				// 2 different pairs
				auto check_pairs = [&text, &length]()
				{
					for (int i = 0; i < length - 2; i++)
					{
						if (text[i] == text[i + 1])
						{
							for (int j = i + 2; j < length - 1; j++)
							{
								if (text[j] == text[j + 1] && text[i] != text[j])
								{
									return true;
								}
							}
						}
					}
					return false;
				};

				if (!check_pairs())
				{
					continue;
				}

				break;
			}
		};

		int length = 0;

		// parse input
		while (*input != '\n')
		{
			length++;
			input++;
		}

		input -= length;

		// copy input
		char* text = new char[length + 1];
		memcpy(text, input, length);
		text[length] = '\0';

		// part 1
		get_next_password(text, length);

		// copy answer to stringResult
		for (int i = 0; i < length; i++)
		{
			stringResult.first[i] = text[i];
		}

		// part 2
		get_next_password(text, length);

		// copy answer to stringResult
		for (int i = 0; i < length; i++)
		{
			stringResult.second[i] = text[i];
		}

		delete[] text;

		return { part1, part2 };
	}
};
